File: TheWorld/wwwroot/js/tripsController.js
﻿// tripsController.js

(function() {

    "user strict";

    // getting the existing module
    angular.module("app-trips")
      .controller("tripsController", tripsController);

    function tripsController($http) {

        var vm = this;

        vm.trips = [];
        vm.newTrip = {};
        vm.errorMessage = "";

        vm.isBusy = true;

        $http.get("/api/trips")
            .then(function(response) {
                    // Success
                    angular.copy(response.data, vm.trips);
                },
                function(error) {
                    // failure
                    vm.errorMessage = "Failed to load trip data: " + error;
                })
            .finally(function() {
                vm.isBusy = false;
            });

        vm.addTrip = function() {

            vm.isBusy = true;
            vm.errorMessage = "";

            $http.post("/api/trips", vm.newTrip)
                .then(function(response) {
                        // success
                        vm.trips.push(response.data);
                        vm.newTrip = {};
                    },
                    function() {
                        vm.errorMessage = "Failed to save new trip data: " + error;

                    })
                .finally(function() {
                    vm.isBusy = false;
                });

        };
    }

})();

File: TheWorld/Controllers/Api/TripsController.cs
using AutoMapper;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using TheWorld.Models;
using TheWorld.ViewModels;

namespace TheWorld.Controllers.Api
{
    [Route("api/trips")]
    [Authorize]
    public class TripsController : Controller
    {
        private readonly IWorldRepository _repository;
        private readonly ILogger<TripsController> _logger;

        public TripsController(IWorldRepository repository, ILogger<TripsController> logger)
        {
            _repository = repository ?? throw new ArgumentNullException(nameof(repository));
            _logger = logger;
        }

        [HttpGet]
        public IActionResult Get()
        {
            try
            {
                var usersTrips = _repository.GetTripsByUsername(this.User.Identity.Name);
                return Ok(Mapper.Map<IEnumerable<TripViewModel>>(usersTrips));
            }
            catch (Exception ex)
            {
                // TODO: Logging
                _logger?.LogError($"Failed to get all trips {ex}");

                return BadRequest("Error occurred");
            }
        }

        [HttpPost]
        public async Task<IActionResult> Post([FromBody] TripViewModel theTrip)
        {
            if (ModelState.IsValid)
            {
                // Save to the Database
                var newTrip = Mapper.Map<Trip>(theTrip);
                newTrip.UserName = User.Identity.Name;

                _repository.AddTrip(newTrip);

                if (await _repository.SaveChangesAsync())
                {
                    return Created($"api/trips/{theTrip.Name}", Mapper.Map<TripViewModel>(newTrip));
                }
            }

            return BadRequest("Failed to save the trip");
        }
    }
}
